File: articles/linear-algebra-for-systems-modeling/articles/inverse-matrices-and-structural-recovery/fortran/inverse_recovery.f90
program inverse_recovery
  implicit none

  real :: a, b, c, d, y1, y2, det, x1, x2

  a = 3.0
  b = 1.0
  c = 2.0
  d = 4.0
  y1 = 7.0
  y2 = 8.0

  det = a * d - b * c

  if (det == 0.0) then
    print *, "Matrix is singular; recovery is not unique."
  else
    x1 = (d * y1 - b * y2) / det
    x2 = (-c * y1 + a * y2) / det
    print *, "Recovered state:"
    print *, "x1 =", x1
    print *, "x2 =", x2
  end if
end program inverse_recovery

File: articles/linear-algebra-for-systems-modeling/articles/inverse-matrices-and-structural-recovery/java/InverseRecovery.java
public class InverseRecovery {
    public static void main(String[] args) {
        double a = 3, b = 1, c = 2, d = 4;
        double y1 = 7, y2 = 8;
        double det = a * d - b * c;

        if (det == 0) {
            System.out.println("Matrix is singular; recovery is not unique.");
            return;
        }

        double x1 = (d * y1 - b * y2) / det;
        double x2 = (-c * y1 + a * y2) / det;

        System.out.printf("Recovered state: x1 = %.2f, x2 = %.2f%n", x1, x2);
    }
}
